docs/flow/config-language/components.md

```diff
@@ -6,3 +6,157 @@ weight: 300
+	detector       = < "fsnotify" | "poll" >
+	poll_frequency = <duration> 
+	*/
+
+	// Exports: a single field named `content`
+	// It can be referred to as `local.file.token.content`
+}
+```
+
+To wire components together, you just use the Exports of one as the Arguments
+of another.
+
+For example, here's a component that scrapes Prometheus metrics. The `targets`
+field is populated with two scrape targets; a constant one `localhost:9001` and
+an expression that ties the target to the contents of
+`local.file.target.content`.
+
+```river
+prometheus.scrape "default" {
+	targets = [
+		{ "__address__" = local.file.target.content }, // tada!
+		{ "__address__" = "localhost:9001" },
+	] 
+
+	forward_to = [prometheus.remote_write.default.receiver]
+	scrape_config {
+		job_name = "default"
+	}
+}
+```
+
+Every time the file contents change and the `local.file` component re-evaluates
+its Exports, the value will propagated to `prometheus.scrape` so it can start
+scraping the new target.
+
+All Arguments and Exports have an underlying [type]({{< relref "./expressions/types_and_values.md" >}}).
+River will type-check expressions before assigning a value to an attribute; the
+documentation of each component will have more information about the ways that
+you can wire components together.
+
+River attributes that appear as Arguments and Exports can carry around more
+than the basic types you'd expect such as strings, numbers or booleans; they
+are able to natively wrap any Go value (eg. channels or interfaces), as well as
+pass around and invoke Go functions which makes for powerful tooling when
+building more complex pipelines.
+
+## DAG
+The relationships between components form a [Directed Acyclic Graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph).
+
+This graph is used by River to resolve dependencies between components, set the
+correct order for component evaluation, and ensure the validity of references.
+River does not allow components that reference themselves, any other
+relationship that might end up in cyclical references, and will reject outright
+invalid references in expressions.
+
+In the previous example, the contents of the `local.file.target.content`
+expression must first be evaluated in a concrete value then type-checked and
+substituted into `prometheus.scrape.default` for it to be configured in turn.
+
+## Component Lifecycle
+Each time a River configuration file is reloaded or re-evaluated, Flow will:
+
+- Parse the provided configuration file. In case any errors are found, the
+  Agent will continue using the last valid River configuration.
+- Parse and apply settings for predefined global blocks (like `logging`).
+- Create a graph of component relationships and validate it.
+- Create new components that weren't present in the last loaded configuration
+  file.
+- Drop components that no longer exist in the new configuration file.
+- Update any pre-existing components whose arguments have changed.
+
+## Error reporting
```

Comment:
I think we can move the enriched error messages to the index file as a generic "this is what River can do" rather than giving specific examples for errors you might get for components, syntax, or expressions.